man/vlock.1
.TH VLOCK 1 "28 July 2007" "Linux" "Linux User's Manual"
.SH NAME
vlock \- Virtual Console lock program
.SH SYNOPSIS
.B vlock [ -hv ]
.PP
.B vlock [ -acns ] [ -t <timeout> ] [ plugins... ]
.SH DESCRIPTION
.B vlock
is a program to lock one or more sessions on the Linux console.  This is
especially useful for Linux machines which have multiple users with access
to the console.  One user may lock his or her session(s) while still allowing
other users to use the system on other virtual consoles.  If desired, the
entire console may be locked and virtual console switching disabled.
.PP
By default, only the current VC (virtual console) is locked.  With the
\fB-a,--all\fR option all VCs are locked.  The locked VCs cannot be unlocked
without the invoker's password or the root password.  The root password will
always be able to unlock any or all sessions, unless disabled at compile time.
.PP
Please note that it is entirely possible to completely lock yourself out of the
console with the \fB-a,--all\fR option if you cannot remember your password!
Unless you are able to kill vlock by logging in remotely via a serial terminal
or network, a hard reset is the only method of ``unlocking'' the display.
.PP
When locking the entire console display it is sometimes still possible to kill
vlock using the Secure Access Key (SAK) or other commands that are available
through the SysRq mechanism.  When the \fB-s,--disable-sysrq\fR and
\fB-a,--all\fR options are given the SysRq mechanism is disabled while vlock is
running.  See /usr/src/linux/Documentation/sysrq.txt for more details.
.PP
vlock works for console sessions primarily.  To lock the entire console display
from an X session use the \fB-n,--new\fR option.  This will make vlock switch
to an empty virtual console to lock the display. 
.PP
The options \fB-n,--new\fR, \fB-s,--disable-sysrq\fR, and \fB-t,--timeout\fR
only work if vlock is compiled with plugin support.  See the PLUGINS section
for more information.
.SH OPTIONS
.B -a,--all
.IP
Lock all console sessions and disable VC switching.
.PP
.B -c,--current
.IP
Lock the current session (this is the default).
.PP
.B -n,--new
.IP
Switch to a new virtual console before locking all console sessions.
.PP
.B -s,--disable-sysrq
.IP
Disable the SysRq mechanism while consoles are locked.  This option only works
if the \fB-a,--all\fR option given.
.PP
.B -t,--timeout <seconds>
.IP
Specify the timeout for the screensaver plugins.  See vlock-plugins(5) for more
information.
.PP
.B -h,--help
.IP
Print a brief help message.
.PP
.B -v,--version
.IP
Print the version number.
.PP
.SH "ENVIRONMENT VARIABLES"
The following environment variables can be used to change vlock's behavior:
.PP
.B USER
.IP
If this variable is when \fBvlock\fR is run as root (uid 0) vlock locks the
screen as this user instead of root.  The root password will still be able to
unlock the session, unless disabled at compile time.
.PP
.B VLOCK_ALL_MESSAGE
.IP
If this variable is set and all consoles are locked its contents will be used
as the locking message instead of the default message.
.PP
.B VLOCK_CURRENT_MESSAGE
.IP
If this variable is set and only the current consoles is locked its contents
will be used as the locking message instead of the default message.
.PP
.B VLOCK_MESSAGE
.IP
If this variable is set its contents will be used as the locking message
instead of the default.  This overrides the former two variables.
.PP
.B VLOCK_PLUGINS
.IP
If this variable is set it is interpreted as a space separated list of plugins
that will be loaded when vlock starts additionally to the ones listed on the
command line.
.PP
.B VLOCK_TIMEOUT
.IP
Set this variable to specify the timeout (in seconds) after which the screen
saver plugins (if any) will be invoked.  If this variable is unset or set to an
invalid value or 0 no timeout is used.  See vlock-plugins(5) for more
information about plugins.
.PP
.B VLOCK_PROMPT_TIMEOUT
.IP
Set this variable to specify the amount of time (in seconds) you will have to
enter your password at the password prompt.  If this variable is unset or set
to an invalid value or 0 no timeout is used.  \fBWarning\fR: If this value is
too low, you may not be able to unlock your session.
.PP
.SH FILES
.B ~/.vlockrc
.IP
This file is read by \fBvlock\fR on startup if it exists.  All the variables
mentioned above can be set here.
.SH SECURITY
See the SECURITY file in the \fBvlock\fR distribution for more information.
.PP
.SH "SEE ALSO"
.BR vlock-main (8),
.BR vlock-plugins (5)
.SH AUTHORS
Michael K. Johnson <http://danlj.org/mkj/>
.PP
Frank Benkstein <frank-vlock@benkstein.net>


ChangeLog
2011-09-18 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 2.2.3

	* Fix vlock not reacting to input when started in the background
	from bash.  (Thanks to Amy Fong for the report.)

2008-05-14 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 2.2.2

	* Fix and improve build system: use LDLIBS instead of LDFLAGS where
	appropriate and provide EXTRA_CFLAGS/EXTRA_LDFLAGS parameters in ./configure.
	* Display number of authentication tries on exit.

2008-03-26 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 2.2.1

	* Fix a script handling bug.

2008-03-21 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 2.2

	* Small plugin documentation update.

2008-01-15 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 2.2 rc5

	* Fixed snprintf() corner case in "new" module.
	* Restore timeout if select() was interrupted by a signal in prompt().

2007-12-29 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 2.2 rc4

	* Slightly changed the fix for the previous security issue: directory parts
	are now silently stripped from erroneous names instead of reported as errors.
	* Fixed a file descriptor leak in script handling.
	* Fixed a bug where vlock would run in an infinite loop if not started from
	a terminal.  Now it simply exits.
	* Added unit tests (using cunit).

2007-11-28 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 2.2 rc3

	* A critical security problem (local root exploit) was fixed:
		Previous versions of vlock (since 2.2 alpha1) contained a serious flaw
		that allowed any user to execute arbitrary code as root .  This was
		possible because plugin names could contain "../" thus escaping the
		pre-defined plugin directory.  All users of vlock 2.2 development versions
		are advised to update.
	* Improved script handling.
	* Added two more scripts.

2007-11-26 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 2.2 rc2

	* Fixed man page installation error.

2007-11-17 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 2.2 rc1

	* Updated documentation.

2007-11-04 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 2.2 beta2

	* Improved error handling and error messages.
	* Fixed portability bugs.
	* Fixed dependency handling bug.

2007-10-16 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 2.2 beta1

	* Documentation update.
	* Improved error handling.

2007-10-03 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 2.2 alpha3

	* Added ./configure script.
	* Improved plugin handling.
	* Improved error handling.
	* Improved child process handling.

2007-09-21 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 2.2 alpha2

	* Added libcaca based screen saver module.
	* Miscellaneous minor fixes.

2007-09-19 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 2.2 alpha1

	* Medium rewrite was undertaken.
	* Support for plugins (modules and scripts) was added.
	* vlock-all, vlock-new, and vlock-nosysrq were converted to modules.
	* Other sample modules and scripts were added.
	* vlock now tries to terminale cleanly if killed by SIGTERM and on errors.

2007-09-08 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 2.1

	* Documentation update.

2007-08-29 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 2.1 rc1

	* Really tiny cleanups.

2007-08-23 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 2.1 beta1

	* Fix compilation on Debian/kFreeBSD.

2007-08-16 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 2.1 alpha2

	* Locking message now configurable.
	* Added configuration file support: ~/.vlockrc.
	* "vlock --new" now works even if stdin is not a terminal.
	* Improved documentation.

2007-08-11 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 2.1 alpha1

	* Added FreeBSD support.  Build with PAM_LIBS=-lpam, install with
	  INSTALL=ginstall.
	* Improved signal handling.
	* Added timeout support to password prompts.

2007-08-09 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 2.0

	* No changes since vlock 2.0 rc2.

2007-08-06 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 2.0 rc2

	* Improved man page and error messages.
	* Fixed bug that prevented shadow authentication from working.
	* vlock can now only be killed by root.  This will probably be
	  fixed in future releases.

2007-08-05 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 2.0 rc1

	* Code cleanups (comments, style).
	* Fixed two merely theoretical security problems.

2007-08-02 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 2.0 beta2

	* Many improvements to security.
	* General cleanup to source tree.
	* ChangeLog added.
	* Manpages updated.
	* LICENSE file updated, license terms unchanged.
	* README updated.
	* Added a SECURITY blurb.
	* -n,--new option now implies -a,--all instead of requiring it.

2007-07-30 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 2.0 beta1

	* Added manpages for all tools.
	* Enter key must be pressed before authentication is started.
	* When run as root, vlock now locks as the user given by the
	  $USER environment variable.
	* Add -n,--new option to run vlock on a new virtual console.

2007-07-28 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 2.0 alpha2

	* Shadow authentication added.

2007-07-25 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 2.0 alpha1

	* Complete rewrite. vlock is now made of several tools called
	  from a shell script.

2007-05-23 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 1.4

	* No changes since rc2.

2007-05-10 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 1.4 rc2

	* Make install rule made more friendly to packagers.
	* Updated license declarations: sources now clearly
	  state "GNU General Public License version 2".

2007-05-09 Frank Benkstein <frank-vlock@benkstein.net>

	vlock 1.4 rc1

	* Some compile time warnings fixed..
	* Support for SysRq disabling added.

1999-01-13 Michael Johnson <johnsonm@redhat.com>

	vlock 1.3

1998-03-12 Michael Johnson <johnsonm@redhat.com>

	vlock 1.2


src/vlock.sh
#!%BOURNE_SHELL%
#
# vlock.sh -- start script for vlock, the VT locking program for linux
# 
# This program is copyright (C) 2007 Frank Benkstein, and is free
# software which is freely distributable under the terms of the
# GNU General Public License version 2, included as the file COPYING in this
# distribution.  It is NOT public domain software, and any
# redistribution not permitted by the GNU General Public License is
# expressly forbidden without prior written permission from
# the author.

# Ignore some signals.
trap : HUP INT QUIT TSTP

# Exit on error.
set -e

# Magic characters to clear the terminal.
CLEAR_SCREEN="`echo -e '\033[H\033[J'`"

# Enter message that is common to different the messages.
VLOCK_ENTER_PROMPT="Please press [ENTER] to unlock."

# Message that is displayed when console switching is disabled.
VLOCK_ALL_MESSAGE="${CLEAR_SCREEN}\
The entire console display is now completely locked.
You will not be able to switch to another virtual console.

${VLOCK_ENTER_PROMPT}"

# Message that is displayed when only the current terminal is locked.
VLOCK_CURRENT_MESSAGE="${CLEAR_SCREEN}\
This TTY is now locked.

${VLOCK_ENTER_PROMPT}"

# Read user settings.
if [ -r "${HOME}/.vlockrc" ] ; then
  . "${HOME}/.vlockrc"
fi

# "Compile" time variables.
VLOCK_MAIN="%PREFIX%/sbin/vlock-main"
VLOCK_VERSION="%VLOCK_VERSION%"
# If set to "y" plugin support is enabled in vlock-main.
VLOCK_ENABLE_PLUGINS="%VLOCK_ENABLE_PLUGINS%"

print_help() {
  echo >&2 "vlock: locks virtual consoles, saving your current session."
  if [ "${VLOCK_ENABLE_PLUGINS}" = "yes" ] ; then
    echo >&2 "Usage: vlock [options] [plugins...]"
  else
    echo >&2 "Usage: vlock [options]"
  fi
  echo >&2 "       Where [options] are any of:"
  echo >&2 "-c or --current: lock only this virtual console, allowing user to"
  echo >&2 "       switch to other virtual consoles."
  echo >&2 "-a or --all: lock all virtual consoles by preventing other users"
  echo >&2 "       from switching virtual consoles."
  if [ "${VLOCK_ENABLE_PLUGINS}" = "yes" ] ; then
    echo >&2 "-n or --new: allocate a new virtual console before locking,"
    echo >&2 "       implies --all."
    echo >&2 "-s or --disable-sysrq: disable SysRq while consoles are locked to"
    echo >&2 "       prevent killing vlock with SAK"
    echo >&2 "-t <seconds> or --timeout <seconds>: run screen saver plugins"
    echo >&2 "       after the given amount of time."
  fi
  echo >&2 "-v or --version: Print the version number of vlock and exit."
  echo >&2 "-h or --help: Print this help message and exit."
}

# Export variables only if they are set.  Some shells create an empty variable
# on export even if it was previously unset.
export_if_set() {
  while [ $# -gt 0 ] ; do
    if ( eval [ "\"\${$1+set}\"" = "set" ] ) ; then
      eval export $1
    fi
    shift
  done
}

main() {
  short_options_with_arguments="t"
  long_options_with_arguments="timeout"

  # Parse command line arguments.
  while [ $# -gt 0 ] ; do
    case "$1" in
      -[!-]?*)
        # Strip "-" to get the list of option characters.
        options="${1#-}"
        shift

        last_option_argument="${options}"
        last_option_index=0

        # If an option character takes an argument all characters after it
        # become the argument if it isn't already the last one.  E.g. if "x"
        # takes an argument "-fooxbar" becomes "-foo -x bar".
        while [ -n "${last_option_argument}" ] ; do
          # Get first option character.
          option="$(expr "${last_option_argument}" : '\(.\)')"
          # Strip it from the list of option characters.
          last_option_argument="${last_option_argument#?}"
          last_option_index=$((${last_option_index} + 1))

          if expr "${short_options_with_arguments}" : "${option}" >/dev/null ; then
            # Prepend "-" plus option character and rest of option string to $@.
            set -- "-${option}" "${last_option_argument}" "$@"

            # Remove all characters after the option character.
            if [ "${last_option_index}" -gt 1 ] ; then
              options="$(expr "${options}" : "\(.\{$((${last_option_index}-1))\}\)")"
            else
              options=""
            fi

            break
          fi
        done

        # Convert clashed arguments like "-foobar" to "-f -o -o -b -a -r".
        while [ -n "${options}" ] ; do
          # Get last option character.
          option="$(expr "${options}" : '.*\(.\)')"
          # Strip it from the list of option characters.
          options="${options%?}"
          # Prepend "-" plus option character to $@.
          set -- "-${option}" "$@"
        done
      ;;
      --?*=?*)
        # Extract option name and argument.
        option="$(expr "x$1" : 'x--\([^=]*\)=.*')"
        option_argument="$(expr "x$1" : 'x--[^=]*=\(.*\)')"
        shift

        compare_options="${long_options_with_arguments}"

        # Find the option in the list of options that take an argument.
        while [ -n "${compare_options}" ] ; do
          compare_option="${compare_options%%,*}"
          compare_options="${compare_options#"${compare_option}"}"
          compare_options="${compare_options#,}"

          if [ "${option}" = "${compare_option}" ] ; then
            set -- "--${option}" "${option_argument}" "$@"
            unset option option_argument
            break
          fi
        done

        if [ -n "${option}" ] ; then
          echo >&2 "$0: option '--${option}' does not allow an argument"
          exit 1
        fi
      ;;
      -a|--all)
        plugins="${plugins} all"
        shift
        ;;
      -c|--current)
        unset plugins
        shift
        ;;
      -n|--new)
        plugins="${plugins} new"
        shift
        ;;
      -s|--disable-sysrq)
        plugins="${plugins} nosysrq"
        shift
        ;;
      -t|--timeout)
        VLOCK_TIMEOUT="$2"
        if ! shift 2 ; then
          echo >&2 "$0: option '$1' requires an argument"
          exit 1
        fi
        ;;
      -h|--help)
       print_help
       exit
       ;;
      -v|--version)
        if [ "${VLOCK_ENABLE_PLUGINS}" = "yes" ] ; then
          echo >&2 "vlock version ${VLOCK_VERSION}"
        else
          echo >&2 "vlock version ${VLOCK_VERSION} (no plugin support)"
        fi
        exit
        ;;
      -[!-]|--?*)
        echo >&1 "$0: unknown option '$1'"
        print_help
        exit 1
      ;;
      --)
        # End of option list.
        shift
        break
        ;;
      *)
        for argument ; do
          if [ "${argument}" = "--" ] ; then
            has_double_dash="yes"
            break
          fi
        done

        if [ -n "${has_double_dash}" ] ; then
          set -- "$@" "$1"
        else
          set -- "$@" -- "$1"
        fi

        shift
        ;;
    esac
  done

  # Export variables for vlock-main.
  export_if_set VLOCK_TIMEOUT VLOCK_PROMPT_TIMEOUT
  export_if_set VLOCK_MESSAGE VLOCK_ALL_MESSAGE VLOCK_CURRENT_MESSAGE

  if [ "${VLOCK_ENABLE_PLUGINS}" = "yes" ] ; then
    exec "${VLOCK_MAIN}" ${plugins} ${VLOCK_PLUGINS} "$@"
  else
    exec "${VLOCK_MAIN}" ${plugins}
  fi
}

main "$@"
